netket/optimizer/qgt/qgt_jacobian_pytree_logic.py

```diff
@@ -0,0 +1,287 @@
+    return tree_cast(res, v)
+
+
+# ==============================================================================
+# the logic above only works for R→R, R→C and holomorphic C→C
+# here the other modes are converted
+
+
+@partial(jax.jit, static_argnums=(0, 4, 5))
+def prepare_centered_oks(
+    apply_fun: Callable,
+    params: PyTree,
+    samples: Array,
+    model_state: Optional[PyTree],
+    mode: str,
+    rescale_shift: bool,
+) -> PyTree:
+    """
+    compute ΔOⱼₖ = Oⱼₖ - ⟨Oₖ⟩ = ∂/∂pₖ ln Ψ(σⱼ) - ⟨∂/∂pₖ ln Ψ⟩
+    divided by √n
+
+    In a somewhat intransparent way this also internally splits all parameters to real
+    in the 'real' and 'complex' modes (for C→R, R&C→R, R&C→C and general C→C) resulting in the respective ΔOⱼₖ
+    which is only compatible with split-to-real pytree vectors
+
+    Args:
+        apply_fun: The forward pass of the Ansatz
+        params : a pytree of parameters p
+        samples : an array of (n in total) batched samples σ
+        model_state: untrained state parameters of the model
+        mode: differentiation mode, must be one of 'real', 'complex', 'holomorphic'
+        rescale_shift: whether scale-invariant regularisation should be used (default: True)
+
+    Returns:
+        if not rescale_shift:
+            a pytree representing the centered jacobian of ln Ψ evaluated at the samples σ, divided by √n;
+            None
+        else:
+            the same pytree, but the entries for each parameter normalised to unit norm;
+            pytree containing the norms that were divided out (same shape as params)
+
+    """
+    # un-batch the samples
+    samples = samples.reshape((-1, samples.shape[-1]))
+
+    # pre-apply the model state
+    def forward_fn(W, σ):
+        return apply_fun({"params": W, **model_state}, σ)
+
+    if mode == "real":
+        split_complex_params = True  # convert C→R and R&C→R to R→R
+        centered_jacobian_fun = centered_jacobian_real_holo
+    elif mode == "complex":
+        split_complex_params = True  # convert C→C and R&C→C to R→C
+        # centered_jacobian_fun = compose(stack_jacobian, centered_jacobian_cplx)
+
+        # avoid converting to complex and then back
+        # by passing around the oks as a tuple of two pytrees representing the real and imag parts
+        centered_jacobian_fun = compose(
+            stack_jacobian_tuple,
+            partial(centered_jacobian_cplx, _build_fn=lambda *x: x),
+        )
+    elif mode == "holomorphic":
+        split_complex_params = False
+        centered_jacobian_fun = centered_jacobian_real_holo
+    else:
+        raise NotImplementedError(
+            'Differentiation mode should be one of "real", "complex", or "holomorphic", got {}'.format(
+                mode
+            )
+        )
+
+    if split_complex_params:
+        # doesn't do anything if the params are already real
+        params, reassemble = tree_to_real(params)
+
+        def f(W, σ):
+            return forward_fn(reassemble(W), σ)
```

Comment:
@inailuig While I see the need for this, It's a bit unfortunate we are forced to do it:
this is splitting complex parameters into 2 real parts, then reassembling them, computing the gradient, than disassembling the gradient.

I doubt jax can optimise through the disassemble/reassemble because he loses structure information there (I might be wrong though).

The best thing would be to think of a different solution.
Maybe we should experiment with complex SOA instead of AOS.
I don't know how hard it would be to implement it though

Reply:
Is there anything in JAX that would allow us to reinterpret-cast a complex array into a float array? We don't particularly care about the order of the parameters, so interleaved real and imag parts would be fine

Reply:
I think @inailuig once wrote an aggressive `vmap` that achieved something like this, but it still copies out the array somewhere else

Reply:
> Maybe we should experiment with complex SOA instead of AOS.
> I don't know how hard it would be to implement it though

I suppose you'd have to define a `ComplexArray` dataclass with two real array leaves, and redefine every complex operation we'd like to use (i.e. all of basic matrix arithmetics and much of the LAX wrapper) to work with it as it would with an `ndarray` of `complex`es.

Reply:
Oh! `jax.lax.bitcast_convert_type` is what i was looking for! I could not find it. It would have made much easier (and faster) to split real and imaginary parts.

Anyhow, those are purely performance improvements, let's leave them for the future.

>I have manually checked a disassembled binary and it seems like the jit compiler can in principle actually optimize away two subsequent calls to tree_to_real and reassemble (still doing one single copy; donate_argnums didn't seem to want work). Maybe not here though when we take the grad ...

Wow. That's quite remarkable.

Reply:
very cool though!